### ISSUE TITLE
Java Double Calculations precision issue

### ISSUE THREAD

**Issue by zzkzzzz** [nus-cs2030/2021-s1#82]
Be careful when you use `double` to do calculation. You may get the result is not what is expected.
For example: 
```
jshell> 4.1-1
$1 ==> 3.0999999999999996

jshell> 5.6+5.8
$2 ==> 11.399999999999999
```
This is because `double` is internally stored as a fraction in binary **not in decimal**. It will be rounded to some number of binary digits.

So,in project 1 I have faced this issue which failed my test cases. You may want to compare start time of two events after you do some calculation like plus 1 or minus 1. You may need to use `BigDecimal` to make sure the precise decimal is correct. 
Or can use `Math.abs(a-b)<0.00001` to compare two `double`.

**Comment by marcuspeh** [comment]
Thanks for the heads up. This doesn't just applies for java but for other language too. I guess that's why our answer need to be rounded off to 2dp for our project.

**Comment by azr124** [comment]
Yes! I took the entire night to solve it. I did use Math.abs(a-b)<0.00001 and now it works already. Thank you soo much!

**Comment by Terence-Sim** [comment]
Yes, this is a standard problem in most programming languages. That is, comparing doubles for equality can fail because of precision issues. The following simple code runs forever, even though the intention is for it to stop when `i` reaches 1.0

```Java
double counter = 1.0;
double i = 0.0;
while(i != counter) {
    System.out.println(i);
    i += 0.1;
}
```

Thus, it is a good idea to use the abstraction:
```Java
boolean closeEnough(double a, double b) {
   double tolerance = 1E-4;
   return Math.abs(a-b) < tolerance;
}
```
